### PR TITLE
More sharing hooks for extended auditing

### PR DESCRIPTION
**Pull request by blizzz**
Two new hooks: share_link_access and post_update_password

and a bugfix, object instances cannot be compared with ===

Please review @rullzer @schiesbn @PVince81

### CODE REVIEW

**Comment by karlitschek** [comment]
looks good :+1:

**Comment by blizzz** [comment]
@rullzer about that failing test, this goes against ShareController::authenticate(), but as far as I can see, it is not used anywhere, except in tests. Can this be removed totally? But I probably oversee something since you worked on this method not long ago?

**Comment by rullzer** [comment]
@blizz will take a look :)

**Comment by rullzer** [comment]
@blizz aaah it is because of the lazy shares... I'll fix it... plus add some test for the hooks.. so we know that they are still thrown if somebody messes with the code.

File: apps/files_sharing/lib/controllers/sharecontroller.php [hunk 1]
```diff
@@ -190,6 +191,45 @@ private function linkShareAuth(\OCP\Share\IShare $share, $password = null) {
...
+	 * @param string $errorMessage
+	 * @throws NotFoundException
+	 * @throws OC\HintException
+	 * @throws OC\ServerNotAvailableException
+	 */
+	protected function emitAccessShareHook($share, $errorCode = 200, $errorMessage = '') {
+		$itemType = $itemSource = $uidOwner = '';
+		$token = $share;
+		$exception = null;
+		if($share instanceof \OC\Share20\Share) {
```
Comment (rullzer) [inline apps/files_sharing/lib/controllers/sharecontroller.php]: use `OCP\Share\IShare`

File: apps/files_sharing/lib/controllers/sharecontroller.php [hunk 2]
```diff
@@ -190,6 +191,45 @@ private function linkShareAuth(\OCP\Share\IShare $share, $password = null) {
 	}
 
 	/**
+	 * throws hooks when a share is attempted to be accessed
+	 *
+	 * @param \OC\Share20\Share|string $share the Share instance if available,
```
Comment (rullzer) [inline apps/files_sharing/lib/controllers/sharecontroller.php]: jikes :S

File: lib/private/share20/manager.php [hunk 3]
```diff
@@ -595,9 +595,17 @@ public function updateShare(\OCP\Share\IShare $share) {
...
+					'itemType' => $share->getNode() instanceof \OCP\Files\File ? 'file' : 'folder',
+					'itemSource' => $share->getNode()->getId(),
+					'uidOwner' => $share->getSharedBy(),
+					'token' => $share->getToken(),
+					'disabled' => is_null($share->getPassword()),
+				]);
 			}
 
-			if ($share->getExpirationDate() !== $originalShare->getExpirationDate()) {
+			if ($share->getExpirationDate() != $originalShare->getExpirationDate()) {
```
Comment (rullzer) [inline lib/private/share20/manager.php]: good catch :)

File: lib/private/share20/manager.php [hunk 4]
```diff
@@ -595,9 +595,17 @@ public function updateShare(\OCP\Share\IShare $share) {
 				if ($share->getPassword() !== null) {
 					$share->setPassword($this->hasher->hash($share->getPassword()));
 				}
+
+				\OC_Hook::emit('OCP\Share', 'post_update_password', [
```
Comment (rullzer) [inline lib/private/share20/manager.php]: No this is wrong! emitting the hook here before things are updated is not correct.

**Comment by rullzer** [comment]
Added some more commits... they can be properly sqaused later

**Comment by PVince81** [comment]
Code looks good :+1:

**Comment by PVince81** [comment]
@rullzer @blizzz can you review each other's commits so we can move this forward ?

**Comment by nickvergessen** [comment]
:+1:

**Comment by blizzz** [comment]
All great, thank you @rullzer :+1:

**Pull request merged**